### ISSUE TITLE
Wondering How to Best build a Retry Policy with a service that returns remaining requests

### ISSUE THREAD

**Issue by drdamour** [App-vNext/Polly#956]
**Summary: What are you wanting to achieve?**
We're consuming a service that instead of offering a retry-after, only has the following headers (on almost all responses)
* X-RateLimit-Limit-minute - how many we can send max per a minute
* X-RateLimit-Remaining-minute - how may we can send in the current minute

We call this from a azure function on a consumption plan so the auto scaling nature of that means many processes and no way to "share" a policy instance across those process boundaries.  I'm wanting to build a policy that uses response info to know when to rate limit and retry things based on these response headers.  It can't be perfect and would definitley need a wrapping retry policy that responded to the eventual 429s that would pop up.

In other words i want a rate policy set to N per second BUT i want to be able to pull tokens out of the bucket and put em back myself based on the X-RateLimit-Remaining-minute response header.

wondering if anyones attempted something like this, or maybe has a drastically different approach from what i'm conceiving of.

**What code or approach do you have so far?**  
nothing of value, this is pretty ugly stuff IMO and kinda lost where to begin

It is _always_ useful to see:

+ **code of any policy declarations**, 
+ **code of the calls to `.Execute/AndCapture/Async(...)`**

**Comment by drdamour** [comment]
Still seeking help

**Comment by martincostello** [comment]
See our retry documentation: https://github.com/App-vNext/Polly/wiki/Retry#retryafter-when-the-response-specifies-how-long-to-wait

**Comment by drdamour** [comment]
@martincostello unless im mistaken that isnt the scenario im asking about

**Comment by martintmk** [comment]
@drdamour 

> In other words i want a rate policy set to N per second BUT i want to be able to pull tokens out of the bucket and put em back myself based on the X-RateLimit-Remaining-minute response header.

In Polly V8 we are using the .NET Rate Limiting API underneath and we allow great customization on how the rate limiter behaves. 

For example:

``` csharp 
var limiter = new TokenBucketRateLimiter(new TokenBucketRateLimiterOptions()
{
    AutoReplenishment = true
});

var pipeline = new ResiliencePipelineBuilder()
    .AddRateLimiter(new RateLimiterStrategyOptions()
    {
        RateLimiter = args =>
        {
            args.Context.Properties.TryGetValue(new ResiliencePropertyKey<string>("some-token"), out var token);
            // do something with token

            // or whatever complex logic you want to use to determine the rate limiter
            return limiter.AcquireAsync(1, args.Context.CancellationToken);
        }
    })
    .Build();
```

Anything you can do with .NET Rate Limiting you can do in Polly.

**Comment by drdamour** [comment]
never could figure out what this answer was actually saying, probably got to read a lot more docs....
